webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/reports/NodeConfigurationService.scala

```diff
@@ -0,0 +1,350 @@
+   * The queue size is 1 and new request need to merge with existing request
+   * It's a List and not a Set, because we want to keep the precedence in
+   * invalidation request.
+   * // unsure if its a CacheComplianceQueueAction or another queueaction
+   */
+  private[this] val invalidateNodeConfigurationRequest = Queue.dropping[List[(NodeId, CacheComplianceQueueAction)]](1).runNow
+
+  /**
+   * We need a semaphore to protect queue content merge-update
+   */
+  private[this] val invalidateMergeUpdateSemaphore = Semaphore.make(1).runNow
+
+
+  // Init to do
+  // what's the best method ? init directly from db, fetching all nodeconfigurations
+  // that are empty
+  // or initing it with all nodes, and nothing in it, and only then fetching by batch
+  // batching saves memory, but is slower
+  // i think it's safer to do the batching part, but i'd like to be sure of that
+  def init() : IOResult[Unit] = {
+    logEffect.debug("Init cache in NodeConfigurationService")
+    for {
+      // first, get all nodes
+      nodeIds <- nodeInfoService.getAllNodesIds()
+      _       <- semaphore.withPermit {
+        // void the cache
+        IOResult.effect({ cache = nodeIds.map(_ -> None).toMap })
+      }
+    } yield ()
+
+  }
+  /**
+   * Update logic. We take message from queue one at a time, and process.
+   * we need to keep order
+   */
+  val updateCacheFromRequest: IO[Nothing, Unit] = invalidateNodeConfigurationRequest.take.flatMap(invalidatedIds =>
+    ZIO.foreach_(invalidatedIds.map(_._2) : List[CacheComplianceQueueAction])(action =>
+    {
+      (for {
+        _  <- performAction(action)
+      } yield ()).catchAll(err => ReportLoggerPure.Cache.error(s"Error when updating NodeConfiguration cache for node: [${action.nodeId.value}]: ${err.fullMsg}"))
+    }
+    )
+  )
+  // start updating
+  updateCacheFromRequest.forever.forkDaemon.runNow
+  /**
+   * Clear cache. Try a reload asynchronously, disregarding
+   * the result
+   */
+  override def clearCache(): Unit = {
+    init().runNow
+    logEffect.debug("Node expected reports cache cleared")
+  }
+
+
+  /**
+   * Do something with the action we received
+   */
+  private[this] def performAction(action: CacheComplianceQueueAction): IOResult[Unit] = {
+    import CacheComplianceQueueAction._
+    // in a semaphore
+    semaphore.withPermit(
+       action match {
+                          case insert: InsertNodeInCache => IOResult.effectNonBlocking {  cache = cache + (insert.nodeId -> None) }
+                          case delete: RemoveNodeInCache => IOResult.effectNonBlocking {  cache = cache.removed(delete.nodeId) }
+                          case update: UpdateNodeConfiguration => IOResult.effectNonBlocking {  cache = cache + (update.nodeId -> Some(update.nodeConfiguration)) }
+                          case something =>
+                            Inconsistency(s"NodeConfiguration service cache received unknown command : ${something}").fail
```

Comment:
I'm going to try to make a subtype for action to be sure that that property holds

Reply:
It works (with some boilerplate), so we can make it checkable: https://github.com/ncharles/rudder/pull/10